File: python/tree/convert-sorted-array-to-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def bcreate(self,start,end):
        if end < start:
            return None

        mid = (start + end) / 2
        node = TreeNode(self.nums[mid])
        node.left = self.bcreate(start,mid -1)
        node.right = self.bcreate(mid + 1, end)

        return node

    def sortedArrayToBST(self, nums):
        self.nums = nums
        return self.bcreate(0,len(self.nums) - 1)

File: c/tree/convert-sorted-array-to-binary-search-tree.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
struct TreeNode *bcreate(int *nums, int start, int end)
{
    if(end < start)
        return NULL;

    int mid = (start + end) / 2;
    struct TreeNode *node = malloc(sizeof(struct TreeNode));

    node->val = nums[mid];
    node->left = bcreate(nums, start, mid - 1);
    node->right = bcreate(nums, mid + 1, end);

    return node;
}

struct TreeNode* sortedArrayToBST(int* nums, int numsSize) {
    return bcreate(nums,0,numsSize - 1);
}
